src/async_wrap.cc

```diff
@@ -311,33 +350,29 @@ static void SetupHooks(const FunctionCallbackInfo<Value>& args) {
   SET_HOOK_FN(promise_resolve);
 #undef SET_HOOK_FN
 
-  {
-    Local<FunctionTemplate> ctor =
-        FunctionTemplate::New(env->isolate());
-    ctor->SetClassName(FIXED_ONE_BYTE_STRING(env->isolate(), "PromiseWrap"));
-    Local<ObjectTemplate> promise_wrap_template = ctor->InstanceTemplate();
-    promise_wrap_template->SetInternalFieldCount(
-        PromiseWrap::kInternalFieldCount);
-    promise_wrap_template->SetAccessor(
-        FIXED_ONE_BYTE_STRING(env->isolate(), "isChainedPromise"),
-        PromiseWrap::getIsChainedPromise);
-    env->set_promise_wrap_template(promise_wrap_template);
-  }
+  PromiseWrap::Initialize(env);
 }
 
-
 static void EnablePromiseHook(const FunctionCallbackInfo<Value>& args) {
-  args.GetIsolate()->SetPromiseHook(PromiseHook);
+  Environment* env = Environment::GetCurrent(args);
+
+  if (args[0]->IsFunction()) {
```

Comment:
What will happen with this new approach in the following scenario?
```js
hookWithoutDestroy.enable();
Promise.resolve(42)
  .then(() => {
    hookWithoutDestroy.disable();
    hookWithDestroy.enable();
  });
```

With the existing implementation, it would still emit `after` and `destroy` events, while it looks like with the latest changes from this PR it won't (`FullPromiseHook` will try to unwrap the promise, but the `wrap` will be null, as the wrap was made on JS side).

So, my question is whether we can make sure that `FastPromiseHook` and `FullPromiseHook` are compatible with each other or not.

Reply:
We discussed this in the diagnostics working group call today and I think we agreed that switching from hooks with a destroy to hooks _with_ a destroy hook is similar to the switch from no hooks so it's reasonable to expect promises created before the switch to not emit the destroy.

Reply:
I haven't tested it, but I suspect the promise would actually just get upgraded due to the existing native promise hook handler building a fresh wrap if `extractPromiseWrap` returns a null pointer.

Reply:
> I think we agreed that switching from hooks with a destroy to hooks _with_ a destroy hook is similar to the switch from no hooks so it's reasonable to expect promises created before the switch to not emit the destroy.

You probably mean switching from hooks without a `destroy` hook to hooks with a `destroy` hook.

But my question is more than that. If we loose the JS wrap, we also loose the original async resource id, so in the discussed scenario we get an `after` event emitted for a completely different async resources (async id are different, as well as wraps), yet they stand for the same Promise.

This probably also applies to parent promises, i.e. once we switch to native `FullPromiseHook` we loose all existing parent JS wraps.

Hopefully, I'm missing something, but if I'm not that's a serious problem.

Reply:
Ah, yep, it would have different ids. I could probably make the PromiseWrap constructor check the promise for the symbols from the JS version and use those, if present. 🤔

Reply:
> I could probably make the PromiseWrap constructor check the promise for the symbols from the JS version and use those, if present.

We need a bidirectional interop, i.e. JS version should be able to read async id generated by the native one. That's because user code might switch from hooks with a `destroy` hook to hooks without a `destroy` hook and vice versa. So, things get complicated with having two (JS and native) Promise hooks.

Reply:
Is the performance improvement here because we move code from native to Js or because the `PromiseWrap` objects have been removed?
If it is because of removing `PromiseWrap` objects we could keep both hook variants native. It's still complicated but at least in one programming language and at one location.
if this doesn't work out it's most likely needed to store/read `triggerAsyncIdSymbol`, `isChainedPromise` and `async_id_symbol` also in the full promise hook.

Reply:
> Is the performance improvement here because we move code from native to Js or because the `PromiseWrap` objects have been removed?

@Flarna My 2 cents. In my understanding the performance improvement may be explained by the removal of `PromiseWrap` objects and GC tracking for the no `destroy` hook case (i.e. the fast track). I'm not sure which of them is more valuable, although.

But as we didn't run benchmarks for the latest code yet, I'm not sure if it improves performance via the fast track. Refer to https://github.com/nodejs/node/pull/32891#issuecomment-616055234 to see benchmark results of the previous version of this PR, which had all promise hook logic on the JS side.

Reply:
@Flarna The move to JS is mostly because my original hope was that I could pull PromiseHook out to an entirely separate API which async_hooks could just consume internally to track promises. Seems to be more difficult than I had hoped though due to the extra responsibilities of the AsyncWrap class like emitting trace events.

The majority of the performance hit of the existing code is indeed in the gc-tracking parts of AsyncWrap, but it's a bit unknown if perhaps the performance would be better making another native class which doesn't descend from AsyncWrap. 🤔

Reply:
This _should_ be fixed, but I'll leave this unresolved for now so others can try it out and to give me time to write some _proper_ tests. 😅

Reply:
As far as I remember the idea on Diag summit was to move also the hook provided by v8 as this would be the key to avoid the C++/JS border at all. But I fear nothing happened at v8 side regarding this.

Reply:
I don't think that's going to change with V8 any time soon. My thinking is that Node.js itself could try to avoid the barrier hop wherever possible for its own async types. I get deeper into the idea in nodejs/diagnostics#376.